mattermost-community/mattermost-plugin-google-drive
# Unit tests

#### Summary
I had to use mocks for the unit tests due to the use of the Google api package. I apologize but the tests are hard to read due to the amount of mocking 😅 

- Added unit tests for most of the api endpoints
- Unit test for the token replenishment job
- Moved some code around so I can mock it: oauth2 library and mutex logic
- Fixed a few bugs I found while writing the unit tests

#### Test cases
1. Notification refresh. We don’t have a test case for this in the spreadsheet but we need to add this. We have a job that runs every 12 hours and it will check if there are any notification channels that are due to expire in 24 hours or less, it will create a new notification channel with Google if that is the case
- /google-drive notifications start
- In your database, find the PluginKeyValueStore table find the entry for your user with the pkey `drive_change_channels-<userid>`. The PValue will looks something like this:
        ```
        {
	    "channel_id": "3376dacc-fc44-456c-83da-e89fb46ebff6",
	    "resource_id": "<resource>",
	    "mm_user_id": "jrhwa8mch38hjymsr7kuy7smnh",
	    "expiration": 1731962203000,
	    "token": "<token>",
	    "page_token": "1053"
        }
        ```
- Change the expiration value so that it’s due to expire in the next 24 hours. You will need to do this in the database.
- Restart the plugin or remove it/re-upload, on startup it will run the notification refresh job
- **Expected Result**: You have a new expiration. You will also have a new token
2. Quick regression test of: 
- connecting your account
- disconnecting your account
- A few notification test cases of creating comments
